tests/UnifiedSpecTests/UnifiedSpecTest.php

```diff
@@ -0,0 +1,346 @@
+    }
+
+    /**
+     * @dataProvider provideFailingTests
+     */
+    public function testFailingTests(...$args)
+    {
+        // Cannot use expectException(), as it ignores PHPUnit Exceptions
+        $failed = false;
+
+        try {
+            $this->testCase(...$args);
+        } catch (Throwable $e) {
+            $failed = true;
+        }
+
+        assertTrue($failed, 'Expected test to throw an exception');
+    }
+
+    public function provideFailingTests()
+    {
+        return $this->provideTests(__DIR__ . '/valid-fail');
+    }
+
+    private function provideTests(string $dir)
+    {
+        $testArgs = [];
+
+        foreach (glob($dir . '/*.json') as $filename) {
+            /* Decode the file through the driver's extended JSON parser to
+             * ensure proper handling of special types. */
+            $json = toPHP(fromJSON(file_get_contents($filename)));
+
+            $description = $json->description;
+            $schemaVersion = $json->schemaVersion;
+            $runOnRequirements = $json->runOnRequirements ?? null;
+            $createEntities = $json->createEntities ?? null;
+            $initialData = $json->initialData ?? null;
+            $tests = $json->tests;
+
+            /* Assertions in data providers do not count towards test assertions
+             * but failures will interrupt the test suite with a warning. */
+            $message = 'Invalid test file: ' . $filename;
+            $this->assertInternalType('string', $description, $message);
+            $this->assertInternalType('string', $schemaVersion, $message);
+            $this->assertInternalType('array', $tests, $message);
+
+            foreach ($json->tests as $test) {
+                $this->assertInternalType('object', $test, $message);
+                $this->assertInternalType('string', $test->description, $message);
+
+                $name = $description . ': ' . $test->description;
+                $testArgs[$name] = [$test, $schemaVersion, $runOnRequirements, $createEntities, $initialData];
+            }
+        }
+
+        return $testArgs;
+    }
+
+    /**
+     * Checks server version and topology requirements.
+     *
+     * @param array $runOnRequirements
+     * @throws SkippedTest unless one or more runOnRequirements are met
+     */
+    private function checkRunOnRequirements(array $runOnRequirements)
+    {
+        $this->assertNotEmpty($runOnRequirements);
+        $this->assertContainsOnly('object', $runOnRequirements);
+
+        $serverVersion = $this->getCachedServerVersion();
+        $topology = $this->getCachedTopology();
+
+        foreach ($runOnRequirements as $o) {
+            $runOnRequirement = new RunOnRequirement($o);
+            if ($runOnRequirement->isSatisfied($serverVersion, $topology)) {
+                return;
+            }
+        }
+
+        $this->markTestSkipped(sprintf('Server version "%s" and topology "%s" do not meet test requirements', $serverVersion, $topology));
```

Comment:
I'll see if I can print the requirements in the message as well. I think it would be helpful, WDYT?

Reply:
I'm used to opening tests based on the name, so up to you. I imagine you'd either want to spread it across multiple lines or `json_encode()` the entire thing for single-line output.

Reply:
In that case, let's keep it as-is for the time being. 👍